conda_forge_tick/update_upstream_versions.py

```diff
@@ -0,0 +1,169 @@
+
+
+class PyPI:
+    name = 'pypi'
+
+    def get_url(self, meta_yaml):
+        url_names = ['pypi.python.org', 'pypi.org', 'pypi.io']
+        source_url = meta_yaml['url']
+        if not any(s in source_url for s in url_names):
+            return None
+        pkg = meta_yaml['url'].split('/')[6]
+        return 'https://pypi.org/pypi/{}/json'.format(pkg)
+
+    def get_version(self, url):
+        r = requests.get(url)
+        if not r.ok:
+            return False
+        return r.json()['info']['version'].strip()
+
+
+class CRAN(LibrariesIO):
+    name = 'cran'
+    url_contains = 'cran.r-project.org/src/contrib/Archive'
+    package_name = lambda self, url: url.split('/')[6]
+    def get_version(self, url):
+        ver = LibrariesIO.get_version(self, url)
+        return str(ver).replace('-', '_')
+
+
+
+class RawURL:
+    name = 'RawURL'
+    def get_url(self, meta_yaml):
+        pkg = meta_yaml['name']
+        url_template = "https://raw.githubusercontent.com/conda-forge/{}-feedstock/master/recipe/meta.yaml"
+        try:
+            content = requests.get(url_template.format(pkg)).content
+            content = content.decode('utf-8')
+        except Exception:
+            return None
+
+        env = jinja2.Environment(undefined=NullUndefined)
+
+        current_ver = meta_yaml['version']
+        orig_ver = current_ver
+        found = True
+        while found:
+            found = False
+            for next_ver in next_version(current_ver):
+                new_content = content.replace(orig_ver, next_ver)
+                meta = parsed_meta_yaml(new_content)
+                url = str(meta['source']['url'])
+                if str(meta['package']['version']) != next_ver:
+                    continue
+                with open(os.devnull, 'w') as devnull:
+                    try:
+                        subprocess.check_call(["wget", "--spider", url], stdout=devnull, stderr=subprocess.STDOUT)
+                    except:
+                        continue
+                    found = True
+                    current_ver = next_ver
+                    break
+        if current_ver != orig_ver:
+            return current_ver
+
+    def get_version(self, url):
+        return url
+
+
+
+def get_latest_version(meta_yaml, sources):
+    for source in sources:
+        url = source.get_url(meta_yaml)
+        if url is None:
+            continue
+        ver = source.get_version(url)
+        if ver:
+            return ver
+        else:
+            with open('upstream_bad', 'a') as f:
+                f.write('{}: Could not find version on {}\n'.format(
+                    meta_yaml['name'], source.name))
+    with open('upstream_bad', 'a') as f:
+        f.write('{}: unknown source\n'.format(meta_yaml['name']))
+    return False
+
+
+def update_upstream_versions(gx, sources=[PyPI(), CRAN(), Github()]):
+
+    for node, attrs in gx.node.items():
+        attrs['new_version'] = get_latest_version(attrs, sources)
```

Comment:
If the version is coming from CRAN would it be possible to flip the version back to something that the CRAN upstream understands? (There were some issues finding the right tarballs last round)